src/crate/client/sqlalchemy/tests/query_caching.py

```diff
@@ -0,0 +1,117 @@
+except ImportError:
+    from sqlalchemy.ext.declarative import declarative_base
+
+from crate.client.sqlalchemy.types import Object, ObjectArray
+
+
+class SqlAlchemyQueryCompilationCaching(TestCase):
+
+    def setUp(self):
+        self.engine = sa.create_engine(f"crate://{crate_host}")
+        self.metadata = sa.MetaData(schema="testdrive")
+        self.session = Session(bind=self.engine)
+        self.Character = self.setup_entity()
+
+    def setup_entity(self):
+        """
+        Define ORM entity.
+        """
+        Base = declarative_base(metadata=self.metadata)
+
+        class Character(Base):
+            __tablename__ = 'characters'
+            name = sa.Column(sa.String, primary_key=True)
+            age = sa.Column(sa.Integer)
+            data = sa.Column(Object)
+            data_list = sa.Column(ObjectArray)
+
+        return Character
+
+    def setup_data(self):
+        """
+        Insert two records into the `characters` table.
+        """
+        self.metadata.drop_all(self.engine)
+        self.metadata.create_all(self.engine)
+
+        Character = self.Character
+        char1 = Character(name='Trillian', data={'x': 1}, data_list=[{'foo': 1, 'bar': 10}])
+        char2 = Character(name='Slartibartfast', data={'y': 2}, data_list=[{'bar': 2}])
+        self.session.add(char1)
+        self.session.add(char2)
+        self.session.commit()
+        self.session.execute(sa.text("REFRESH TABLE testdrive.characters;"))
+
+    @skipIf(SA_VERSION < SA_1_4, "On SA13, the 'ResultProxy' object has no attribute 'scalar_one'")
+    def test_object_multiple_select(self):
+        """
+        The SQLAlchemy implementation of CrateDB's `OBJECT` type offers indexed
+        access to the instance's content in form of a dictionary. Thus, it must
+        not use `cache_ok = True` on its implementation, i.e. this part of the
+        compiled SQL clause must not be cached.
+
+        This test verifies that two subsequent `SELECT` statements are translated
+        well, and don't trip on incorrect SQL compiled statement caching.
+        """
+        self.setup_data()
+        Character = self.Character
+
+        selectable = sa.select(Character).where(Character.data['x'] == 1)
+        result = self.session.execute(selectable).scalar_one().data
+        self.assertEqual({"x": 1}, result)
+
+        selectable = sa.select(Character).where(Character.data['y'] == 2)
+        result = self.session.execute(selectable).scalar_one().data
+        self.assertEqual({"y": 2}, result)
+
+    @skipIf(SA_VERSION < SA_1_4, "On SA13, the 'ResultProxy' object has no attribute 'scalar_one'")
+    def test_objectarray_multiple_select(self):
+        """
+        The SQLAlchemy implementation of CrateDB's `ARRAY` type in form of the
+        `ObjectArray`, does *not* offer indexed access to the instance's content.
+        Thus, using `cache_ok = True` on that type should be sane, and not mess
```

Comment:
but now we disable cache for both cases, no? and it's not configurable by the user?

Reply:
Caching of compiled statements is only turned off for `OBJECT` types, implemented on behalf of `_Craty`. See https://github.com/crate/crate-python/pull/559#pullrequestreview-1512359974.

Reply:
Apologies if I'm still confused, but `Character` is still a custom type, I don't get when this cache is used and when not.

Reply:
> SQLAlchemy as of version 1.4 includes a [SQL compilation caching facility](https://docs.sqlalchemy.org/en/20/core/connections.html#sql-caching) which will allow Core and ORM SQL constructs to cache their stringified form, along with other structural information used to fetch results from the statement, allowing the relatively expensive string compilation process to be skipped when another structurally equivalent construct is next used. This system relies upon functionality that is implemented for all SQL constructs, including objects such as [Column](https://docs.sqlalchemy.org/en/20/core/metadata.html#sqlalchemy.schema.Column), [select()](https://docs.sqlalchemy.org/en/20/core/selectable.html#sqlalchemy.sql.expression.select), and [TypeEngine](https://docs.sqlalchemy.org/en/20/core/type_api.html#sqlalchemy.types.TypeEngine) objects, to produce a cache key which fully represents their state to the degree that it affects the SQL compilation process.

-- https://docs.sqlalchemy.org/en/20/faq/performance.html ff.

Reply:
Currently, accessing the `OBJECT` implementation `_Craty` using dict-like indexing breaks that machinery.